scalding-core/src/main/scala/com/twitter/scalding/typed/TypedPipe.scala

```diff
@@ -471,7 +471,18 @@ trait TypedPipe[+T] extends Serializable {
    * Reasonably common shortcut for cases of total associative/commutative reduction
    * returns a ValuePipe with only one element if there is any input, otherwise EmptyValue.
    */
-  def sum[U >: T](implicit plus: Semigroup[U]): ValuePipe[U] = ComputedValue(groupAll.sum[U].values)
+  def sum[U >: T](implicit plus: Semigroup[U]): ValuePipe[U] = {
+    // every 1000 items, compact.
+    lazy implicit val batchedSG = Batched.compactingSemigroup[U](1000)
+    ComputedValue(map { t => ((), Batched[U](t)) }
+      .sumByLocalKeys
+      // remove the Batched before going to the reducers
+      .map { case (_, batched) => batched.sum }
```

Comment:
Couldn't this be expressed in a `StatefulSummer` instead (and wouldn't that maybe be a better place for it)? If so, it could be combined with `AdaptiveCache` to ensure that we don't run out of memory. And we already have config there (around the cache size and/or using the adaptive one), so this would be a natural extension of that.